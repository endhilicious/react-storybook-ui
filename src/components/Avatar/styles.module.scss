.Avatar__image {
  border-radius: 50%;
  object-fit: cover;
}
.Avatar__nameOnly {
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 50%;
  background-color: #E7EAF4;
  border: 2px solid #fff;
  font-weight: 500;
  font-size: 14px;
  line-height: 20px;
  color: #767B87;
  border-radius: 50%;
}