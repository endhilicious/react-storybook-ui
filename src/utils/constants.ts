export const DEFAULT_AVATAR_NAME = 'Ooo Ooo';
export const ALT_IMAGE = 'https://st3.depositphotos.com/1767687/16607/v/600/depositphotos_166074422-stock-illustration-default-avatar-profile-icon-grey.jpg';
export const MOCK_IMAGES = [
  'https://assets.materialup.com/uploads/b78ca002-cd6c-4f84-befb-c09dd9261025/preview.png',
  'https://assets.materialup.com/uploads/b78ca002-cd6c-4f84-befb-c09dd9261025/preview.png',
  'https://assets.materialup.com/uploads/b78ca002-cd6c-4f84-befb-c09dd9261025/preview.png',
  'https://assets.materialup.com/uploads/b78ca002-cd6c-4f84-befb-c09dd9261025/preview.png',
];
export const MAX_LIMIT_AVATAR = 3;
export const MOCK_IMAGE = 'https://assets.materialup.com/uploads/b78ca002-cd6c-4f84-befb-c09dd9261025/preview.png';