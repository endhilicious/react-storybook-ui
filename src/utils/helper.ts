export const getSize = (size: string): number => {
  switch (size) {
    case 'xs':
      return 24;
    case 'md':
      return 48;
    case 'lg':
      return 64;
    case 'sm':
    default:
      return 32;
  }
};

export const getAvatarTextSize = (size: string): number => {
  switch (size) {
    case 'xs':
      return 8;
    case 'md':
      return 20;
    case 'lg':
      return 24;
    case 'sm':
    default:
      return 14;
  }
};

export const getAbbreviatedName = (name: string): string | null => {
  if (!name) return null;
  const maxWords = 2;
  const cleanedName = name.replace(/\s+\.\s+/g, ' ');
  const nameWords = cleanedName.split(' ');
  const abbreviation = nameWords
    .slice(0, maxWords)
    .map((word) => word.charAt(0))
    .join('');
  return abbreviation;
}
