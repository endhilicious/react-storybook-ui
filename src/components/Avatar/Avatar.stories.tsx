import { ComponentStory, ComponentMeta } from '@storybook/react';
import { MOCK_IMAGE } from '../../utils/constants';

import Avatar from './Avatar';

export default {
  title: 'Example/Avatar',
  component: Avatar,
  parameters: {
    layout: 'fullscreen',
  },
} as ComponentMeta<typeof Avatar>;

const Template: ComponentStory<typeof Avatar> = (args) => (
  <div style={{ padding: 14 }}>
    <Avatar {...args} />
  </div>
);

export const ExtraSmall = Template.bind({});
ExtraSmall.args = {
  src: MOCK_IMAGE,
  size: 'xs',
};

export const Small = Template.bind({});
Small.args = {
  src: MOCK_IMAGE,
  size: 'sm',
};

export const Medium = Template.bind({});
Medium.args = {
  src: MOCK_IMAGE,
  size: 'md',
};

export const Large = Template.bind({});
Large.args = {
  src: MOCK_IMAGE,
  size: 'lg',
};
