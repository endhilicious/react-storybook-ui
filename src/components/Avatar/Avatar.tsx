import { FC } from 'react';
import cx from 'classnames';
import { Size } from '../../utils/globalTypes';
import { getAbbreviatedName, getAvatarTextSize, getSize } from '../../utils/helper';

import styles from './styles.module.scss';
import { ALT_IMAGE, DEFAULT_AVATAR_NAME } from '../../utils/constants';


interface AvatarProps {
  classNames?: string;
  avatarName?: string;
  hideImage?: boolean;
  src: string;
  size: Size;
}

const Avatar: FC<AvatarProps> = ({
  src,
  size,
  classNames = '',
  avatarName = DEFAULT_AVATAR_NAME,
  hideImage = false 
}) => {
  if (hideImage)
    return (
      <div
        className={cx(styles.Avatar__nameOnly, classNames)}
        style={{ width: getSize(size), height: getSize(size) }}
      >
        <span style={{ fontSize: getAvatarTextSize(size) }}>
          {getAbbreviatedName(avatarName)}
        </span>
      </div>
    )

  return (
    <img
      src={src}
      alt={src}
      className={cx(styles.Avatar__image, classNames)}
      width={getSize(size)}
      height={getSize(size)}
      onError={({ currentTarget }) => {
        currentTarget.onerror = null;
        currentTarget.src = ALT_IMAGE;
      }}
    />
  );
};

export default Avatar;
