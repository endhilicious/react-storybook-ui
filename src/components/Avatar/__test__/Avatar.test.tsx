import { render, screen } from '@testing-library/react';
import { MOCK_IMAGE } from '../../../utils/constants';
import Avatar from '../Avatar';

describe('Avatar', () => {

  it('renders the default component properly', () => {
    render(<Avatar src={MOCK_IMAGE} size="md" />);
    const avatarElement = screen.getByAltText(MOCK_IMAGE);
  
    expect(avatarElement).toBeInTheDocument();
    expect(avatarElement).toHaveAttribute('src', MOCK_IMAGE);
    expect(avatarElement).toHaveClass('Avatar__image');
    expect(avatarElement).toHaveAttribute('width', '48');
    expect(avatarElement).toHaveAttribute('height', '48');
  });
  

  it('renders avatars with custom size', () => {
    render(<Avatar src={MOCK_IMAGE} size="sm" />);
    const avatarElement = screen.getByAltText(MOCK_IMAGE);

    expect(avatarElement).toHaveAttribute('width', '32');
    expect(avatarElement).toHaveAttribute('height', '32');
  });

  it('renders avatars with additional classNames', () => {
    render(<Avatar src={MOCK_IMAGE} size="md" classNames="custom-class" />);
    const avatarElement = screen.getByAltText(MOCK_IMAGE);

    expect(avatarElement).toHaveClass('Avatar__image');
    expect(avatarElement).toHaveClass('custom-class');
  });

  it('renders the name-only component when hideImage prop is true', () => {
    render(<Avatar src={MOCK_IMAGE} size="md" hideImage avatarName="Oktavianus Octavianus" />);
    const nameOnlyElement = screen.getByText('OO');
  
    expect(nameOnlyElement).toBeInTheDocument();
  });
  

  it('captures a component snapshot file', () => {
    const { asFragment } = render(<Avatar src={MOCK_IMAGE} size="md" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
