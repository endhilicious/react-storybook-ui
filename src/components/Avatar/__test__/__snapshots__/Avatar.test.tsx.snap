// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar captures a component snapshot file 1`] = `
<DocumentFragment>
  <img
    alt="https://assets.materialup.com/uploads/b78ca002-cd6c-4f84-befb-c09dd9261025/preview.png"
    class="Avatar__image"
    height="48"
    src="https://assets.materialup.com/uploads/b78ca002-cd6c-4f84-befb-c09dd9261025/preview.png"
    width="48"
  />
</DocumentFragment>
`;
